src/tss2-tcti/tcti-tbs.c

```diff
@@ -0,0 +1,295 @@
+/* SPDX-License-Identifier: BSD-2 */
+/*
+ * Copyright (c) 2015 - 2018 Intel Corporation
+ * All rights reserved.
+ */
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+#include <windows.h>
+#include <Tbs.h>
+
+#include "tss2_tcti.h"
+#include "tss2_tcti_tbs.h"
+
+#include "tcti-common.h"
+#include "tcti-tbs.h"
+
+#define LOGMODULE tcti
+#include "util/log.h"
+
+/*
+ * This function wraps the "up-cast" of the opaque TCTI context type to the
+ * type for the TBS TCTI context. The only safe-guard we have to ensure
+ * this operation is possible is the magic number for the TBS TCTI context.
+ * If passed a NULL context, or the magic number check fails, this function
+ * will return NULL.
+ */
+TSS2_TCTI_TBS_CONTEXT*
+tcti_tbs_context_cast (TSS2_TCTI_CONTEXT *tcti_ctx)
+{
+    if (tcti_ctx != NULL && TSS2_TCTI_MAGIC (tcti_ctx) == TCTI_TBS_MAGIC) {
+        return (TSS2_TCTI_TBS_CONTEXT*)tcti_ctx;
+    }
+    return NULL;
+}
+
+/*
+ * This function down-casts the TBS TCTI context to the common context
+ * defined in the tcti-common module.
+ */
+TSS2_TCTI_COMMON_CONTEXT*
+tcti_tbs_down_cast (TSS2_TCTI_TBS_CONTEXT *tcti_tbs)
+{
+    if (tcti_tbs == NULL) {
+        return NULL;
+    }
+    return &tcti_tbs->common;
+}
+
+TSS2_RC
+tcti_tbs_transmit (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t command_size,
+    const uint8_t *command_buffer)
+{
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs = tcti_tbs_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_tbs_down_cast (tcti_tbs);
+    TSS2_RC rc = TSS2_RC_SUCCESS;
+    TBS_RESULT tbs_rc;
+    /* Reset resSize */
+    tcti_tbs->resSize = sizeof (BYTE) * TBS_RESULT_MAX_BUFFER_SIZE;
+
+    if (tcti_tbs == NULL) {
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    }
+    rc = tcti_common_transmit_checks (tcti_common, command_buffer);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+    LOGBLOB_DEBUG (command_buffer,
+                   command_size,
+                   "sending %zu byte command buffer:",
+                   command_size);
+    tbs_rc = Tbsip_Submit_Command (tcti_tbs->hContext,
```

Comment:
What are your thoughts on making a separate `commandBuffer` to use for this purpose, instead of using one buffer for both the commands and results? Not sure if its worth the extra memory usage, but I think it would improve the readability and might ease the troubleshooting of any future issues.

Reply:
Any reason we can't pass the caller provided `response_buffer` directly to the `Tbsip_Submit_Command` function call when it's moved to the `tcti_tbs_receive` function? If so then we'll only need a single buffer that will hold the contents of the caller provided `command_buffer` in `tcti_tbs_transmit` and then passed to `Tbsip_Submit_Command` as the `pabCommand` parameter.

Reply:
I think this could cause issues when a null `response_buffer` is passed into `tcti_tbs_receive`. In this case `tcti_tbs_receive` is expected to return the size of the response, which is stored in `resSize` by `Tbsip_Sumbit_Command`. In this case `Tbsip_Submit_Command` would be called with the `pabResult`  parameter as null since `response_buffer` is null, and the response would be lost.

Reply:
Where is the Tbsip_Submit_Command() defined?

Reply:
`Tbsip_Submit_Command` is part of Microsoft's TPM Base Services, the documentation for `Tbsip_Submit_Command` can be found [here](https://docs.microsoft.com/en-us/windows/desktop/api/tbs/nf-tbs-tbsip_submit_command).

Reply:
I hadn't considered that case @davidjmaria. I'm still of the mind that the blocking call needs to be `receive`  to keep us from straying too far from the intended behavior. To allow callers to get the size of the response buffer using the `receive` function will require the TCTI maintain / allocate buffers to hold both the command and response then. This is horribly inefficient but seems to be a necessary side effect of wrapping synchronous function in an API designed for an asynchronous one. We should consider this as a use-case when we next update the TCTI API 😒

Reply:
The other option would be to have `tcti_tbs_receive` behave similarly to `tcti_device_receive`, which just returns the max result buffer size whenever `response_buffer` is null, instaed of returning the actual size of the response.

Reply:
Works for me. Like you say, I've taken that route in the past.

Reply:
@flihp Are you aware of any maximum command size defined in the specs? I haven't found anything, in which case I'm going to have to have the transmit and receive functions manage the command buffer instead of having one large command buffer that is managed in context create and finalize.

Reply:
The spec only doesn't define it, but it practice both command and response can't be bigger than 4K. TSS defines both to 4K:
https://github.com/tpm2-software/tpm2-tss/blob/master/include/tss2/tss2_tpm2_types.h#L18
so does the reference TPM implementation from Microsoft:
https://github.com/Microsoft/ms-tpm-20-ref/blob/master/TPMCmd/tpm/include/Implementation.h#L196

Reply:
Since this was a big change, I added it as a separate commit to make review of it easier. I plan on squashing that commit after the review.